README.md
# Bump and Release

![npm](https://img.shields.io/npm/v/bump-and-release?label=version&logo=npm)
![David](https://img.shields.io/david/aversini/bump-and-release?logo=npm)
![David](https://img.shields.io/david/dev/aversini/bump-and-release?logo=npm)
![GitHub Workflow Status](https://img.shields.io/github/workflow/status/aversini/bump-and-release/coverage?label=coverage&logo=github)

> Bump and Release is a tiny command line helper to bump and release packages.

## Installation

```sh
> cd your-project
> npm install --save-dev bump-and-release
```

## Configuration

Update your `package.json` file with the following scripts calls:

```json
"scripts": {
  "bump": "bump-and-release -t bump",
  "release": "bump-and-release-t release"
}
```

If you are using a configuration file, you can pass it via the parameter `-c`:

```json
"scripts": {
  "bump": "bump-and-release -t bump -c ./config/app-config.js",
  "release": "bump-and-release-t release -c ./config/app-config.js"
}
```

Or if you name it `.bump-and-release.config.js` and put it in the same location as your `package.json` file, it will be taken into account automatically.

This:

```json
  "bump": "bump-and-release -t bump -c .bump-and-release.config.js"
```

is equivalent to:

```json
  "bump": "bump-and-release -t bump"
```

if the file `.bump-and-release.config.js` can be found next to your `package.json`.

## Bump

By default, the "bump" command will:

1. Run the following preflight validations
   - check if the current branch name is `master`
   - check if the current branch has a tracking remote
   - check if that remote is either `github/master` or `origin/master`
   - check if your repo has uncommitted files
1. Prompt the user to choose from a list of possible "new versions"
   - patch
   - minor
   - major
1. Update your `package.json` file with the new version
1. Update your `package-lock.json` file with the new version
1. Stage and commit those files
1. Push to the tracking remote

If any of those tasks fail, the process stops.

### Bump example

<img height="451" width="810" src="https://raw.githubusercontent.com/aversini/bump-and-release/master/configuration/bump.png" alt="bump example">

## Release

By default, the "release" command will:

1. Run the following preflight validations
   - check if the current branch name is `master`
   - check if the current branch has a tracking remote
   - check if that remote is `github/master`or `origin/master`
   - check if your repo has uncommitted files
1. Stage and commit any modified files
1. Tag the version with the following template: `vCurrentVersion`
1. Push code and tags to the tracking remote

If any of those tasks fail, the process stops.

### Release example

<img height="451" width="810" src="https://raw.githubusercontent.com/aversini/bump-and-release/master/configuration/release.png" alt="release example">

## Custom configuration

While the defaults are pretty straight forward, it is possible to override them with a configuration file.

### Example

```js
module.exports = {
  /**
   * { array of strings }
   * List of all the branches this tool is NOT allowed to work with.
   * This entry takes precedence over the next one (allowedBranches).
   * This is ignored if bump.local or release.local is true.
   */
  disallowedBranches: [],
  /**
   * { array of strings }
   * List of all the branches this tool is allowed to work with.
   * This is ignored if bump.local or release.local is true.
   */
  allowedBranches: ["master", "main"],
  /**
   * { array of strings }
   * List of all the remotes this tool is allowed to work with.
   * This is ignored if bump.local or release.local is true.
   */
  allowedRemotes: ["github/master", "upstream/main"],

  /**
   * Configurations specific to the "bump" command.
   */
  bump: {
    /**
     * { boolean }
     * Flag indicating that this is a monorepo handled by Lerna with
     * non-independent version (all packages have the same version).
     * Bump will read lerna.json to find the current version, and the
     * location of all the packages listed in the prop "packages".
     * Bump will try to update ALL the packages.json under each and
     * every package listed in that prop.
     */
    lerna: true,
    /**
     * { boolean }
     * Flag indicating that all changes will remain
     * local, preventing anything to be pushed to the remote(s).
     */
    local: true,
    /**
     * { function }
     * Function that takes the new version as a parameter
     * and that returns the string that will be used during
     * the commit phase of the bump.
     */
    commitMessage: (version) =>
      `chore: bumping version for next release: ${version}`,
    /**
     * { array of objects }
     * List of possible "next" versions to be offered to
     * the user at the prompt. These values are the same
     * that are accepted from method `inc()` of the
     * node-semver library.
     * https://github.com/npm/node-semver
     * There is also an extra one called "custom", allowing
     * the user to enter a version manually.
     *
     * NOTE: there is a "special" case which is not a version.
     * It's used to add a separator when all the values are
     * displayed at the prompt.
     * The value for the type is "separator".
     */
    nextPossible: [
      {
        /**
         * { number }
         * The position of this prompt within other prompts.
         * The indexing starts at 0.
         */
        pos: 0,
        /**
         * { string }
         * The type of release as defined in node-semver.
         * Possible values are major, premajor, minor,
         * preminor, patch, prepatch or prerelease.
         * It also accepts "separator".
         */
        type: "prepatch",
        /**
         * { string }
         * Identifier to be used to prefix premajor,
         * preminor, prerelease or prepatch version increments.
         * For example: "alpha", "beta", "rc", etc.
         */
        identifier: "beta",
        /**
         * { boolean }
         * Flag to enable or disable a specific release.
         */
        enabled: true,
        /**
         * { boolean }
         * Flag to make that particular version the default
         * selection.
         */
        default: true,
        /**
         * { function }
         * Function that returns the prompted line for a
         * specific version.
         */
        prompt: (type, version) =>
          `[${type}] ... bump to next ${type} (${version})`,
        /**
         * { number }
         * The desired position for this specific release in
         * the prompt.
         */
      },
      {
        type: "patch",
        prompt: (type, version) =>
          `[${type}] ... bump to next ${type} (${version})`,
      },
      {
        type: "minor",
        prompt: (type, version) =>
          `[${type}] ... bump to next ${type} (${version})`,
      },
      {
        type: "major",
        prompt: (type, version) =>
          `[${type}] ... bump to next ${type} (${version})`,
      },
      {
        type: "separator",
      },
      {
        type: "custom",
        prompt: (type) => `[${type}] ... enter a custom version`,
      },
    ],
    /**
     * { array of objects }
     * List of tasks to run before the bump is actually
     * executed. It is useful to run some tests, or
     * linting one last time for example. Each object has 3
     * keys: the name of the task, the action to run and
     * and optional verbose boolean to indicate if the
     * output of the task should be visible or not (it is
     * not verbose by default).
     *
     * NOTES: any files modified during these steps will
     * automatically be staged, committed and pushed to the
     * remote.
     */
    prebump: [
      {
        name: "run tests",
        command: "npm run test",
        verbose: true,
      },
      {
        name: "lint the files",
        command: "npm run lint",
      },
    ],
  },

  /**
   * Configurations specific to the "release" command.
   */
  release: {
    /**
     * { boolean }
     * Flag indicating that all changes will remain
     * local, preventing anything to be pushed.
     */
    local: true,
    /**
     * { function }
     * Function that takes the current version as a parameter
     * and that returns the string that will be used during
     * the commit phase of the release.
     */
    commitMessage: (version) => `chore: tagging release ${version}`,
    /**
     * { object }
     * Object specifying if this release should be tagged
     * (using the "enabled" boolean flag) and what prefix
     * to use when actually tagging (using the "prefix" string
     * flag).
     */
    tag: {
      enabled: true,
      prefix: "v",
    },
    /**
     * { array of objects }
     * List of tasks to run before the release is actually
     * tagged. It is useful to run some tests, or
     * generate a changelog for example. Each object has 3
     * keys: the name of the task, the action to run and
     * and optional verbose boolean to indicate if the
     * output of the task should be visible or not (it is
     * not verbose by default).
     *
     * NOTES: any files modified during these steps will
     * automatically be staged, committed and pushed to the
     * remote.
     */
    prerelease: [
      {
        name: "run tests",
        command: "npm run test",
        verbose: true,
      },
      {
        name: "generate changelog",
        command: "npm run changelog",
      },
    ],
  },
};
```

### Next Available Versions

The possible "next" versions to be offered to the user at the prompt are the same that are accepted from method `inc()` of the [node-semver library](https://github.com/npm/node-semver).

Here are possible values, assuming the current version of a package is 1.0.0.

| type     | current version | new version  |
| -------- | --------------- | ------------ |
| prepatch | 1.0.0           | 1.0.1-beta.0 |
| patch    | 1.0.0           | 1.0.1        |
| preminor | 1.0.0           | 1.1.0-beta.0 |
| minor    | 1.0.0           | 1.1.0        |
| premajor | 1.0.0           | 2.0.0-beta.0 |
| major    | 1.0.0           | 2.0.0        |

**NOTE**: the "beta" identifier can be customized by providing the key "identifier".

## License

MIT © Arno Versini


CHANGELOG.md
## [3.0.5](https://github.com/aversini/bump-and-release/compare/v3.0.4...v3.0.5) (2023-05-05)


### Bug Fixes

* OMG ([039e880](https://github.com/aversini/bump-and-release/commit/039e88062e18b960a6064e0677e4cf1315491c13))



## [3.0.4](https://github.com/aversini/bump-and-release/compare/v3.0.3...v3.0.4) (2023-05-05)


### Bug Fixes

* still battling with the doggie ([a5acd31](https://github.com/aversini/bump-and-release/commit/a5acd3183c322845f52ee758e387cd1b40f4d7c1))



## [3.0.3](https://github.com/aversini/bump-and-release/compare/v3.0.2...v3.0.3) (2023-05-05)


### Bug Fixes

* still working on husky ([8623296](https://github.com/aversini/bump-and-release/commit/8623296ed41d265abb1406f82c3e0a40150d34af))



## [3.0.2](https://github.com/aversini/bump-and-release/compare/v3.0.1...v3.0.2) (2023-05-05)


### Bug Fixes

* still trying to bypass husky on publish ([2db0e2e](https://github.com/aversini/bump-and-release/commit/2db0e2e5cf0c12c05569b9c2db7c5596d2e6f8e0))



## [3.0.1](https://github.com/aversini/bump-and-release/compare/v3.0.0...v3.0.1) (2023-05-05)


### Bug Fixes

* update husky and prettier to latest ([f5cc863](https://github.com/aversini/bump-and-release/commit/f5cc863b7165d1a4c53fbaef61267cad330ef959))



# [3.0.0](https://github.com/aversini/bump-and-release/compare/v2.7.1...v3.0.0) (2023-05-05)

### Breaking Changes

* Updating dependencies and dropping support for Node v14 and lower.


## [2.7.1](https://github.com/aversini/bump-and-release/compare/v2.7.0...v2.7.1) (2022-06-16)


### Bug Fixes

* if lockfileVersion > 1 updated version in lock file is incomplete ([52a62d7](https://github.com/aversini/bump-and-release/commit/52a62d72fb3290a2d9c77bde33b510a64018e369))



# [2.7.0](https://github.com/aversini/bump-and-release/compare/v2.6.0...v2.7.0) (2022-05-26)


### Features

* ignore branch and remote restriction if local is true ([6e98a65](https://github.com/aversini/bump-and-release/commit/6e98a654b18570a4ba83a4afdf57f3ed41216d5b))



# [2.6.0](https://github.com/aversini/bump-and-release/compare/v2.5.0...v2.6.0) (2022-04-20)


### Features

* adding support for "disallowedBranches" ([66b150d](https://github.com/aversini/bump-and-release/commit/66b150d3b9fe9820f0f12666483cc131a04fe6f3))



# [2.5.0](https://github.com/aversini/bump-and-release/compare/v2.4.1...v2.5.0) (2022-04-14)


### Features

* adding version to custom commands in release ([8c0378f](https://github.com/aversini/bump-and-release/commit/8c0378f5525279a447855c83a1a5bc228b3c9247))



## [2.4.1](https://github.com/aversini/bump-and-release/compare/v2.4.0...v2.4.1) (2022-02-21)


### Bug Fixes

* forgot to pass newversion to prebump action if function ([ae11062](https://github.com/aversini/bump-and-release/commit/ae11062d129df85a0273f859acf3d4bc72101a31))



# [2.4.0](https://github.com/aversini/bump-and-release/compare/v2.3.1...v2.4.0) (2022-02-21)


### Features

* adding support for prebump and prerelease commands to be functions ([54a1979](https://github.com/aversini/bump-and-release/commit/54a1979404b6627314f78a1d46e9874c0473a200))



## [2.3.1](https://github.com/aversini/bump-and-release/compare/v2.3.0...v2.3.1) (2021-11-24)


### Bug Fixes

* adding lerna.json to files that need to be updated ([0ce0c71](https://github.com/aversini/bump-and-release/commit/0ce0c71316a3a756ae6f6bda53bfb882dd34c04d))



# [2.3.0](https://github.com/aversini/bump-and-release/compare/v2.2.1...v2.3.0) (2021-11-24)


### Features

* adding bare minimum support for monorepos with Lerna ([719f410](https://github.com/aversini/bump-and-release/commit/719f4106e38da2ec891ea4f2cb2d562c1e52a293))



## [2.2.1](https://github.com/aversini/bump-and-release/compare/v2.2.0...v2.2.1) (2021-11-24)



# [2.2.0](https://github.com/aversini/bump-and-release/compare/v2.1.0...v2.2.0) (2021-11-24)


### Features

* allowing to pass the version at the CLI ([77a8f25](https://github.com/aversini/bump-and-release/commit/77a8f25bddaf8974d9f10bdcd53a495164691f4c))



# [2.1.0](https://github.com/aversini/bump-and-release/compare/v2.0.1...v2.1.0) (2021-05-20)


### Features

* adding support for "prebump" tasks ([e2f1db3](https://github.com/aversini/bump-and-release/commit/e2f1db3f89b68869aa000e34eefc3e4dd4cdfc01))



## [2.0.1](https://github.com/aversini/bump-and-release/compare/v2.0.0...v2.0.1) (2021-03-05)


### Bug Fixes

* **bump:** better error message when branches have no tracking remote ([835c2c9](https://github.com/aversini/bump-and-release/commit/835c2c9ba4cb77c4a0c352d8b0031eee57940639))



# [2.0.0](https://github.com/aversini/bump-and-release/compare/v1.2.3...v2.0.0) (2021-02-24)


### Features

* locking the default bump version order and removing the automatic separator ([6453831](https://github.com/aversini/bump-and-release/commit/645383178a92649429af651125c48810425a7ddc))


### BREAKING CHANGES

* the separator is not displayed automatically anymore. It needs to be specifically requested in the configuration file via the type="separator" - see the README file for an example



## [1.2.3](https://github.com/aversini/bump-and-release/compare/v1.2.2...v1.2.3) (2021-02-23)


### Bug Fixes

* option "-c" is not required ([f9afc18](https://github.com/aversini/bump-and-release/commit/f9afc18d4ca5478cc906ee79c81f45fd803b272a))



## [1.2.2](https://github.com/aversini/bump-and-release/compare/v1.2.1...v1.2.2) (2021-02-13)


### Bug Fixes

* removing redundant flag (-a) in the commit command ([c40db62](https://github.com/aversini/bump-and-release/commit/c40db62310b10f16519dc9c445afbcc1a890958e))



## [1.2.1](https://github.com/aversini/bump-and-release/compare/v1.2.0...v1.2.1) (2021-02-13)


### Bug Fixes

* bumping to teeny-js-utilities to fix git commit commands potential failures ([da9a8a3](https://github.com/aversini/bump-and-release/commit/da9a8a32f456de683b10245a2a46de6bb4f50deb))

src/utilities.js
const _ = require("lodash");
const inquirer = require("inquirer");
const { red, yellow, cyan } = require("kleur");
const memoizeOne = require("async-memoize-one");
const fs = require("fs-extra");
const path = require("path");
const semver = require("semver");
const boxen = require("boxen");
const TeenyLogger = require("teeny-logger");
const {
  displayErrorMessages,
  runCommand,
  shallowMerge,
} = require("teeny-js-utilities");

const logger = new TeenyLogger({
  boring: process.env.NODE_ENV === "test",
});

const pkg = path.join(process.cwd(), "package.json");
const lernaPkg = path.join(process.cwd(), "lerna.json");

const COMMIT_MESSAGE = "git commit";
const PUSH_MESSAGE = "push";
const BUMP_TYPE_CUSTOM = "custom";
const BUMP_TYPE_SEPARATOR = "separator";
const DEFAULT_BUMP_POSITION = {
  custom: 16,
  major: 15,
  minor: 13,
  patch: 11,
  premajor: 14,
  preminor: 12,
  prepatch: 10,
};

const customizer = (def, cust, key) => {
  if (key === "nextPossible") {
    return _.orderBy(
      _.values(_.merge(_.keyBy(def, "type"), _.keyBy(cust, "type"))),
      ["pos"]
    );
  }
};

const mergeConfigurations = (defaultConfig, customConfig) => {
  const enhancedNextPossible = [];
  const res = shallowMerge(defaultConfig, customConfig, customizer);

  res.bump.nextPossible.forEach((nextPossible) => {
    if (typeof nextPossible.pos === "undefined") {
      nextPossible.pos = DEFAULT_BUMP_POSITION[nextPossible.type];
    }
    enhancedNextPossible.push(nextPossible);
  });
  res.bump.nextPossible = _.orderBy(enhancedNextPossible, ["pos"]);
  return res;
};

const displayConfirmation = async (msg) => {
  const questions = {
    default: true,
    message: "Do you want to continue?",
    name: "goodToGo",
    type: "confirm",
  };

  logger.log();
  logger.log(msg);
  logger.log();

  const answers = await inquirer.prompt(questions);
  return answers.goodToGo;
};

const displayIntroductionMessage = ({ version, branch, remote }) => {
  const dryRunMsg = global["dry-run"]
    ? `\n${yellow("Dry-run mode is ON")}`
    : "";
  const msg = `Current version is ${cyan(version)}
Current branch is ${cyan(branch)}
Current tracking remote is ${cyan(remote)}${dryRunMsg}`;

  logger.log();
  logger.log(
    process.env.NODE_ENV === "test"
      ? msg
      : /* istanbul ignore next */
        boxen(msg, {
          align: "center",
          borderColor: "yellow",
          padding: 1,
        })
  );
  logger.log();
};

const shouldContinue = (goodToGo) => {
  if (!goodToGo) {
    logger.log("\nBye then!");
    process.exit(0);
  }
  return true;
};

const readPackageJSON = async () => {
  let packageJson;
  try {
    packageJson = await fs.readJSON(pkg);
    return packageJson;
  } catch (err) {
    /* istanbul ignore next */
    throw new Error(red(`Unable to parse package.json\n${err}`));
  }
};
const memoizedPackageJSON = memoizeOne(readPackageJSON);

/* istanbul ignore next */
const readLernaJSON = async () => {
  let lernaJson;
  try {
    lernaJson = await fs.readJSON(lernaPkg);
    return lernaJson;
  } catch (err) {
    throw new Error(red(`Unable to parse lerna.json\n${err}`));
  }
};
const memoizedLernaJSON = memoizeOne(readLernaJSON);

const getCurrentVersion = async (config) => {
  /* istanbul ignore next */
  const { version } = config.bump.lerna
    ? await memoizedLernaJSON()
    : await memoizedPackageJSON();
  return version;
};

/* istanbul ignore next */
const getPackagesLocation = async (config) => {
  if (config.bump.lerna) {
    const { packages } = await memoizedLernaJSON();
    return packages;
  }
  return [];
};

const preflightValidation = async (config, local) => {
  /*
   * Check if the current branch is allowed
   * Check if the current branch has a tracking remote
   * Check if the repo is dirty (uncommitted files)
   */
  const branch = await runCommand("git rev-parse --abbrev-ref HEAD", {
    ignoreError: true,
  });
  const remote = await runCommand(
    "git rev-parse --symbolic-full-name --abbrev-ref @{u}",
    { ignoreError: true }
  );
  const dirty = await runCommand("git diff --no-ext-diff --quiet --exit-code", {
    ignoreError: true,
  });
  const version = await getCurrentVersion(config);
  /* istanbul ignore next */
  const packages = config.bump.lerna ? await getPackagesLocation(config) : [];

  if (!local) {
    const errorMessage = [];
    if (config.disallowedBranches.includes(branch)) {
      errorMessage.push(
        `Working branch cannot be one of "${config.disallowedBranches}".`
      );
    } else if (!config.allowedBranches.includes(branch)) {
      errorMessage.push(
        `Working branch must be one of "${config.allowedBranches}".`
      );
    }

    if (!config.allowedRemotes.includes(remote)) {
      errorMessage.push(
        `Tracking remote must be one of "${config.allowedRemotes}".`
      );
    }
    /* istanbul ignore next */
    if (dirty.exitCode > 0) {
      errorMessage.push(
        `Working dir must be clean (no uncommited files).\n${dirty.shortMessage}`
      );
    }
    errorMessage.length && displayErrorMessages(errorMessage);
  }

  return {
    branch,
    packages,
    remote,
    version,
  };
};

const getNextPossibleVersions = ({ current, config }) => {
  const choices = [];
  let index = 0,
    defaultChoice = 0;
  config.bump.nextPossible.forEach((next) => {
    if (next.enabled || typeof next.enabled === "undefined") {
      if (next.default) {
        defaultChoice = index;
      }
      index++;
      const nextVersion =
        next.type !== BUMP_TYPE_CUSTOM && next.type !== BUMP_TYPE_SEPARATOR
          ? semver.inc(current, next.type, next.identifier)
          : next.type;

      if (next.type === BUMP_TYPE_SEPARATOR) {
        choices.push(new inquirer.Separator());
      } else {
        choices.push({
          name: next.prompt
            ? next.prompt(next.type, nextVersion)
            : `[${next.type}] ... bump to ${nextVersion}`,
          short: next.type,
          value: nextVersion,
        });
      }
    }
  });
  return { choices, defaultChoice };
};

const prepareReleaseTasks = (config, version) => {
  const tasks = config.release.prerelease;
  const commands = [];
  const names = [];

  if (tasks.length) {
    tasks.forEach((task) => {
      const name = task.name ? task.name : task.command;
      commands.push({
        action: task.command,
        "dry-run": global["dry-run"],
        name,
        verbose: task.verbose ? task.verbose : false,
      });
      names.push(name);
    });
  }

  const stageMsg = "git stage";
  commands.push({
    action: `git add -A`,
    "dry-run": global["dry-run"],
    name: stageMsg,
  });
  names.push(stageMsg);

  commands.push({
    action: `git commit -m "${config.release.commitMessage(version)}"`,
    "dry-run": global["dry-run"],
    name: COMMIT_MESSAGE,
  });
  names.push(COMMIT_MESSAGE);

  const tagTask = config.release.tag;
  if (tagTask.enabled) {
    const name = "tag";
    commands.push({
      action: `git tag -a ${tagTask.prefix}${version} -m "version ${version}"`,
      "dry-run": global["dry-run"],
      name,
    });
    names.push(name);
  }

  if (!config.release.local) {
    const action = tagTask.enabled
      ? "git push --no-verify && git push --tags --no-verify"
      : "git push --no-verify";
    commands.push({
      action,
      "dry-run": global["dry-run"],
      name: PUSH_MESSAGE,
    });
    names.push(PUSH_MESSAGE);
  }

  return {
    commands,
    // eslint-disable-next-line no-useless-concat
    instruction: `${names.join(", ").replace(/,([^,]*)$/, " and" + "$1")}...`,
  };
};

const prepareBumpTasks = (config, version) => {
  const tasks = config.bump.prebump;
  const commands = [];
  const names = [];

  if (tasks.length) {
    tasks.forEach((task) => {
      const name = task.name ? task.name : task.command;
      commands.push({
        action: task.command,
        "dry-run": global["dry-run"],
        name,
        verbose: task.verbose ? task.verbose : false,
      });
      names.push(name);
    });
  }

  const stageMsg = "git stage & commit";
  commands.push({
    action: `git add -A && git commit -a -m "${config.bump.commitMessage(
      version
    )}"`,
    "dry-run": global["dry-run"],
    name: stageMsg,
  });
  names.push(stageMsg);

  if (!config.bump.local) {
    commands.push({
      action: `git push --no-verify`,
      "dry-run": global["dry-run"],
      name: PUSH_MESSAGE,
    });
    names.push(PUSH_MESSAGE);
  }

  return {
    commands,
    // eslint-disable-next-line no-useless-concat
    instruction: `${names.join(", ").replace(/,([^,]*)$/, " and" + "$1")}...`,
  };
};

module.exports = {
  // private methods
  _getCurrentVersion: getCurrentVersion,
  // constants
  BUMP_TYPE_CUSTOM,
  COMMIT_MESSAGE,
  PUSH_MESSAGE,
  // public methods
  displayConfirmation,
  displayIntroductionMessage,
  getNextPossibleVersions,
  lernaPkg,
  logger,
  memoizedLernaJSON,
  memoizedPackageJSON,
  mergeConfigurations,
  pkg,
  preflightValidation,
  prepareBumpTasks,
  prepareReleaseTasks,
  shouldContinue,
};
